File: csharp/UnitTestProject1/Base64Tests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Base64CSharp;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Base64CSharp.Tests
{
    using BitString = List<int>;
    using Entry = Tuple<int, char>;
    using Dict = List<Tuple<int, char>>;
    [TestClass()]
    public class Base64Tests
    {
        Base64 base64 = new Base64();
        [TestMethod()]
        public void CreateDictionaryTest()
        {
            Assert.AreEqual(base64[0], 'A');
            Assert.AreEqual(base64[1], 'B');
            Assert.AreEqual(base64[25], 'Z');
            Assert.AreEqual(base64[26], 'a');
            Assert.AreEqual(base64[51], 'z');
            Assert.AreEqual(base64[52], '0');
            Assert.AreEqual(base64[62], '+');
            Assert.AreEqual(base64[63], '/');
            Assert.AreEqual(base64.DictLenght, 64);
        }

        [TestMethod()]
        public void padTest()
        {
            string str = "asdasdpoi";
            Assert.AreEqual("asdasdpoiAAA", Base64.pad("right", 'A', 12, str));
            Assert.AreEqual("AAAasdasdpoi", Base64.pad("left", 'A', 12, str));
            Assert.AreEqual(null, Base64.pad("", 'A', 12, str));
        }

        [TestMethod()]
        public void toBitStringTest()
        {
            BitString bits = new BitString() { 1, 1 };
            CollectionAssert.AreEqual(bits, Base64.toBitString(3));
            bits = new BitString() { 1 };
            CollectionAssert.AreEqual(bits, Base64.toBitString(1));
            bits = new BitString();
            CollectionAssert.AreEqual(bits, Base64.toBitString(0));
            bits = new BitString() { 1, 0, 1, 0 };
            CollectionAssert.AreEqual(bits, Base64.toBitString(10));
            bits = new BitString() { 1, 1, 0, 0, 1, 0, 0 };
            CollectionAssert.AreEqual(bits, Base64.toBitString(100));
        }

        [TestMethod()]
        public void fromBitStringTest()
        {
            BitString bits = new BitString() { 0, 0, 0, 0, 0, 0, 1, 1 };
            Assert.AreEqual(Base64.fromBitString(bits), 3);
            bits = new BitString() { 1, 0, 1, 0, 0 };
            Assert.AreEqual(Base64.fromBitString(bits), 20);
            bits = new BitString() { 1, 0, 1, 1, 0 };
            Assert.AreEqual(Base64.fromBitString(bits), 22);
        }

        [TestMethod()]
        public void toBinaryTest()
        {
            string str1 = "s";
            BitString bits = new BitString() { 0, 1, 1, 1, 0, 0, 1, 1 };
            CollectionAssert.AreEqual(bits, Base64.toBinary(str1));
            str1 = "SOS";
            bits = new BitString() { 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1 };
            CollectionAssert.AreEqual(bits, Base64.toBinary(str1));
        }

        [TestMethod()]
        public void chunksOfTest()
        {
            // Test 1
            List<BitString> chunks = new List<BitString> { new BitString() { 1, 2, 3, 4, 5, 6 }, new BitString() { 7, 8, 9, 10 } };
            List<int> chunkTest = Enumerable.Range(1, 10).ToList();
            var chunkz1 = Base64.chunksOf(6, chunkTest);
            if (chunkz1.Count != chunks.Count)
                Assert.Fail();
            else
                for (int i = 0; i < chunks.Count; i++)
                    CollectionAssert.AreEqual(chunks[i], chunkz1[i]);

            // Test 2
            chunks = new List<BitString> { new BitString() { 1, 2, 3 },
                                           new BitString() { 4, 5, 6 },
                                           new BitString() { 7, 8, 9 },
                                           new BitString() { 10, 11, 12 },
                                           new BitString() { 13, 14, 15} };
            chunkTest = Enumerable.Range(1, 15).ToList();
            chunkz1 = Base64.chunksOf(3, chunkTest);
            if (chunkz1.Count != chunks.Count)
                Assert.Fail();
            else
                for (int i = 0; i < chunks.Count; i++)
                    CollectionAssert.AreEqual(chunks[i], chunkz1[i]);
        }

        [TestMethod()]
        public void findCharTest()
        {
            char character = Base64.findChar(base64.GetDict, new BitString() { });
            Assert.AreEqual('A', character);
            character = Base64.findChar(base64.GetDict, new BitString() { 1, 0, 0 });
            Assert.AreEqual('E', character);
            character = Base64.findChar(base64.GetDict, new BitString() { 1, 1, 1, 1, 0, 0 });
            Assert.AreEqual('8', character);
            character = Base64.findChar(base64.GetDict, new BitString() { 1, 1, 1, 1, 1, 1 });
            Assert.AreEqual('/', character);

        }

        [TestMethod()]
        public void translateTest()
        {
            Assert.AreEqual("TWFu", Base64.translate(base64.GetDict, "Man"));
            Assert.AreEqual("cGxlYXN1cmUu", Base64.translate(base64.GetDict, "pleasure."));
            Assert.AreEqual("cGxlYXN1cmU", Base64.translate(base64.GetDict, "pleasure"));
            Assert.AreEqual("cGxlYXN1cg", Base64.translate(base64.GetDict, "pleasur"));
            Assert.AreEqual("cA", Base64.translate(base64.GetDict, "p"));
        }

        [TestMethod()]
        public void encodeTest()
        {
            Assert.AreEqual("U2F2ZSBvdXIgc291bHMh", Base64.encode(base64.GetDict, "Save our souls!"));
            Assert.AreEqual("U2F2ZSBvdXIgc291bHM=", Base64.encode(base64.GetDict, "Save our souls"));
            Assert.AreEqual("U2F2ZSBvdXIgc291bA==", Base64.encode(base64.GetDict, "Save our soul"));
            Assert.AreEqual("U2F2ZSBvdXIgc291", Base64.encode(base64.GetDict, "Save our sou"));
        }

        [TestMethod()]
        public void findCodeTest()
        {
            BitString bits = new BitString() { 1, 1, 0, 1, 0 };
            CollectionAssert.AreEqual(bits, Base64.findCode(base64.GetDict, 'a'));
            bits = new BitString() { 1, 1, 0, 0, 1 };
            CollectionAssert.AreEqual(bits, Base64.findCode(base64.GetDict, 'Z'));
            bits = new BitString() { 1, 1, 1, 1, 0, 1 };
            CollectionAssert.AreEqual(bits, Base64.findCode(base64.GetDict, '9'));
        }

        [TestMethod()]
        public void decodeTest()
        {
            Dict dict = base64.GetDict;
            Assert.AreEqual("Save Our Souls!", Base64.decode(dict, Base64.encode(dict, "Save Our Souls!")));
            Assert.AreEqual("Save Our Souls", Base64.decode(dict, Base64.encode(dict, "Save Our Souls")));
            Assert.AreEqual("Save Our Soul", Base64.decode(dict, Base64.encode(dict, "Save Our Soul")));
            Assert.AreEqual("Base64 is a group of", Base64.decode(dict, Base64.encode(dict, "Base64 is a group of")));
        }
    }
}

File: swift/Base64Swift/Tests/Base64Tests/Base64Tests.swift
import XCTest
@testable import Base64

final class Base64Tests: XCTestCase {
    func testCreateDict() {
        let dict = createDict()
        XCTAssertEqual(Entry(num: 0, c: "A"), dict[0])
        XCTAssertEqual(Entry(num: 1, c: "B"), dict[1])
        XCTAssertEqual(Entry(num: 25, c: "Z"), dict[25])
        XCTAssertEqual(Entry(num: 26, c: "a"), dict[26])
        XCTAssertEqual(Entry(num: 51, c: "z"), dict[51])
        XCTAssertEqual(Entry(num: 52, c: "0"), dict[52])
        XCTAssertEqual(Entry(num: 62, c: "+"), dict[62])
        XCTAssertEqual(Entry(num: 63, c: "/"), dict[63])
    }

    func testPadding() {
        XCTAssertEqual("ASD111", padLeft(input_string: "ASD", filler: "1", len: 6))
        XCTAssertEqual("111ASD", padRight(input_string: "ASD", filler: "1", len: 6))
    }

    func testToBitString() {
        XCTAssertEqual(BitString(), toBitString(num: 0))
        XCTAssertEqual(BitString([1, 0, 1, 0, 0]), toBitString(num: 20))
        XCTAssertEqual(BitString([1, 0, 1, 1, 0]), toBitString(num: 22))
    }

    func testFromBitString() {
        XCTAssertEqual(0, fromBitString(bitString: BitString([])))
        XCTAssertEqual(20, fromBitString(bitString: BitString([1, 0, 1, 0, 0])))
        XCTAssertEqual(22, fromBitString(bitString: BitString([1, 0, 1, 1, 0])))
    }

    func testToBinary() {
        XCTAssertEqual(BitString([0, 1, 1, 1, 0, 0, 1, 1]), toBinary(inputString: "s"))
        XCTAssertEqual(BitString([0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1]), toBinary(inputString: "SOS"))
    }

    func testChunksOf() {
        let z: [[Int]] = []
        XCTAssertEqual(z, chunksOf(num: 8, arr: []))
        let a: [[Int]] = [[1, 2, 3, 4, 5, 6], [7, 8, 9, 10]]
        XCTAssertEqual(a, chunksOf(num: 6, arr: Array(1...10)))
        let b: [[Int]] = [[1,2,3], [4,5,6], [7,8,9], [10, 11, 12], [13, 14, 15]]
        XCTAssertEqual(b, chunksOf(num: 3, arr: Array(1...15)))
    }

    func testFindChar() {
        let dict = createDict()
        XCTAssertEqual("A", findChar(dict: dict, bitString: []))
        XCTAssertEqual("E", findChar(dict: dict, bitString: BitString([1,0,0])))
        XCTAssertEqual("8", findChar(dict: dict, bitString: BitString([1,1,1,1,0,0])))
        XCTAssertEqual("/", findChar(dict: dict, bitString: BitString([1,1,1,1,1,1])))
    }
    
    func testTranslate() {
        let dict = createDict()
        XCTAssertEqual("TWFu", translate(dict: dict, str: "Man"))
        XCTAssertEqual("cGxlYXN1cmUu", translate(dict: dict, str: "pleasure."))
        XCTAssertEqual("cGxlYXN1cmU", translate(dict: dict, str: "pleasure"))
        XCTAssertEqual("cA", translate(dict: dict, str: "p"))
        XCTAssertEqual("", translate(dict: dict, str: ""))
    }

    func testEncode() {
        let dict = createDict()
        XCTAssertEqual("U2F2ZSBvdXIgc291bHMh", encode(dict: dict, str: "Save our souls!"))
        XCTAssertEqual("U2F2ZSBvdXIgc291bHM=", encode(dict: dict, str: "Save our souls"))
        XCTAssertEqual("U2F2ZSBvdXIgc291bA==", encode(dict: dict, str: "Save our soul"))
        XCTAssertEqual("U2F2ZSBvdXIgc291", encode(dict: dict, str: "Save our sou"))
    }

    func testFindCode() {
        let dict = createDict()
        XCTAssertEqual("011010", findCode(dict: dict, c: "a"))
        XCTAssertEqual("011001", findCode(dict: dict, c: "Z"))
        XCTAssertEqual("111101", findCode(dict: dict, c: "9"))
    }

    func testDecode() {
        let dict = createDict()
        XCTAssertEqual("Save our souls!", decode(dict: dict, str: encode(dict: dict, str: "Save our souls!")))
        XCTAssertEqual("Save our souls", decode(dict: dict, str: encode(dict: dict, str: "Save our souls")))
        XCTAssertEqual("Save our soul", decode(dict: dict, str: encode(dict: dict, str: "Save our soul")))
    }


    static var allTests = [
        ("testCreateDict", testCreateDict),
        ("testPadding", testPadding),
        ("testToBitString", testToBitString),
        ("testFromBitString", testFromBitString),
        ("testToBinary", testToBinary),
        ("testChunksOf", testChunksOf),
        ("testFindChar", testFindChar),
        ("testTranslate", testTranslate),
        ("testEncode", testEncode),
        ("testFindCode", testFindCode),
        ("testDecode", testDecode)
    ]
}
